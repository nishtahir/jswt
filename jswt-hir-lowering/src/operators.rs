use jswt_ast::{
    transform::{walk_binary_expression, TransformVisitor},
    ArgumentsExpression, ArgumentsList, BinaryExpression, BinaryOperator, MemberDotExpression,
    SingleExpression,
};
use jswt_common::{Spannable};
use jswt_synthetic::ident_exp;

pub struct HirOperatorsLoweringContext {}

impl HirOperatorsLoweringContext {
    pub fn new() -> Self {
        Self {}
    }
}

impl TransformVisitor for HirOperatorsLoweringContext {
    fn visit_binary_expression(&mut self, node: &BinaryExpression) -> SingleExpression {
        let left = self.visit_single_expression(&node.left);
        let right = self.visit_single_expression(&node.right);
        let ident = match node.op {
            // Arithmetic operators
            BinaryOperator::Plus(_) => "plus",
            BinaryOperator::Minus(_) => "minus",
            BinaryOperator::Mult(_) => "mult",
            BinaryOperator::Div(_) => "div",
            BinaryOperator::Equal(_) => "eq",
            // Comparison operators
            BinaryOperator::NotEqual(_)
            | BinaryOperator::Greater(_)
            | BinaryOperator::GreaterEqual(_)
            | BinaryOperator::Less(_)
            | BinaryOperator::LessEqual(_)
            | BinaryOperator::And(_)
            | BinaryOperator::Or(_) => {
                // TODO implement comparison operators
                return walk_binary_expression(self, node);
            }
            // Assignment operators
            BinaryOperator::Assign(_) => {
                // We can't rewrite assignment expressions into function calls
                return walk_binary_expression(self, node);
            }
        };

        SingleExpression::Arguments(ArgumentsExpression {
            span: node.span(),
            arguments: ArgumentsList {
                span: node.span(),
                arguments: vec![right],
            },
            ident: Box::new(SingleExpression::MemberDot(MemberDotExpression {
                span: node.span(),
                target: Box::new(left),
                expression: Box::new(ident_exp(ident.into(), node.span())),
            })),
        })
    }
}

#[cfg(test)]
mod test {

    use jswt_assert::assert_debug_snapshot;
    use jswt_parser::Parser;
    use jswt_semantics::GlobalSemanticResolver;
    use jswt_symbols::SemanticEnvironment;
    use jswt_tokenizer::Tokenizer;

    use crate::HirLoweringContext;

    #[test]
    fn test_operator_lowering_lowers_plus_symbol_into_function_call() {
        let mut tokenizer = Tokenizer::default();
        tokenizer.enqueue_source_str(
            "test_operator_lowering_lowers_plus_symbol_into_function_call",
            r"
        class i32 {
            add(other: i32): i32 {
                return 0;
            }
        }

        function main(): i32 {
            let a: i32 = 1;
            let b: i32 = 2;
            let x = a + b;
            return 0;
        }
    ",
        );

        let ast = Parser::new(&mut tokenizer).parse();

        let mut environment = SemanticEnvironment::default();
        let mut global_resolver = GlobalSemanticResolver::new(&mut environment);
        global_resolver.resolve(&ast);

        // No errors in global resolver
        assert!(
            global_resolver.errors().len() == 0,
            "{:?}",
            global_resolver.errors()
        );

        let mut lowering = HirLoweringContext::new(&environment);
        let lowered = lowering.lower(&ast);
        assert_debug_snapshot!(lowered);
    }
}
