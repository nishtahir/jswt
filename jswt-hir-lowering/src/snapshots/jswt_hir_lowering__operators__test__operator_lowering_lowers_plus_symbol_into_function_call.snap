---
source: jswt-hir-lowering/src/operators.rs
expression: lowered
---
Ast {
    program: Program {
        files: [
            File {
                span: Span {
                    file: "test_operator_lowering_lowers_plus_symbol_into_function_call",
                    module: "module/test_operator_lowering_lowers_plus_symbol_into_function_call",
                    start: 9,
                    end: 257,
                },
                source_elements: SourceElements {
                    span: Span {
                        file: "test_operator_lowering_lowers_plus_symbol_into_function_call",
                        module: "module/test_operator_lowering_lowers_plus_symbol_into_function_call",
                        start: 9,
                        end: 257,
                    },
                    source_elements: [
                        ClassDeclaration(
                            ClassDeclarationElement {
                                span: Span {
                                    file: "test_operator_lowering_lowers_plus_symbol_into_function_call",
                                    module: "module/test_operator_lowering_lowers_plus_symbol_into_function_call",
                                    start: 9,
                                    end: 105,
                                },
                                annotations: [],
                                export: false,
                                ident: Identifier {
                                    span: Span {
                                        file: "test_operator_lowering_lowers_plus_symbol_into_function_call",
                                        module: "module/test_operator_lowering_lowers_plus_symbol_into_function_call",
                                        start: 15,
                                        end: 18,
                                    },
                                    value: "i32",
                                },
                                body: ClassBody {
                                    span: Span {
                                        file: "test_operator_lowering_lowers_plus_symbol_into_function_call",
                                        module: "module/test_operator_lowering_lowers_plus_symbol_into_function_call",
                                        start: 19,
                                        end: 105,
                                    },
                                    class_elements: [
                                        Method(
                                            ClassMethodElement {
                                                span: Span {
                                                    file: "test_operator_lowering_lowers_plus_symbol_into_function_call",
                                                    module: "module/test_operator_lowering_lowers_plus_symbol_into_function_call",
                                                    start: 33,
                                                    end: 95,
                                                },
                                                annotations: [],
                                                ident: Identifier {
                                                    span: Span {
                                                        file: "test_operator_lowering_lowers_plus_symbol_into_function_call",
                                                        module: "module/test_operator_lowering_lowers_plus_symbol_into_function_call",
                                                        start: 33,
                                                        end: 36,
                                                    },
                                                    value: "add",
                                                },
                                                params: FormalParameterList {
                                                    span: Span {
                                                        file: "test_operator_lowering_lowers_plus_symbol_into_function_call",
                                                        module: "module/test_operator_lowering_lowers_plus_symbol_into_function_call",
                                                        start: 36,
                                                        end: 48,
                                                    },
                                                    parameters: [
                                                        FormalParameterArg {
                                                            span: Span {
                                                                file: "test_operator_lowering_lowers_plus_symbol_into_function_call",
                                                                module: "module/test_operator_lowering_lowers_plus_symbol_into_function_call",
                                                                start: 37,
                                                                end: 47,
                                                            },
                                                            ident: Identifier {
                                                                span: Span {
                                                                    file: "test_operator_lowering_lowers_plus_symbol_into_function_call",
                                                                    module: "module/test_operator_lowering_lowers_plus_symbol_into_function_call",
                                                                    start: 37,
                                                                    end: 42,
                                                                },
                                                                value: "other",
                                                            },
                                                            type_annotation: TypeAnnotation {
                                                                span: Span {
                                                                    file: "test_operator_lowering_lowers_plus_symbol_into_function_call",
                                                                    module: "module/test_operator_lowering_lowers_plus_symbol_into_function_call",
                                                                    start: 44,
                                                                    end: 47,
                                                                },
                                                                ty: Binding(
                                                                    "i32",
                                                                ),
                                                            },
                                                        },
                                                    ],
                                                },
                                                returns: Some(
                                                    TypeAnnotation {
                                                        span: Span {
                                                            file: "test_operator_lowering_lowers_plus_symbol_into_function_call",
                                                            module: "module/test_operator_lowering_lowers_plus_symbol_into_function_call",
                                                            start: 50,
                                                            end: 53,
                                                        },
                                                        ty: Binding(
                                                            "i32",
                                                        ),
                                                    },
                                                ),
                                                body: BlockStatement {
                                                    span: Span {
                                                        file: "test_operator_lowering_lowers_plus_symbol_into_function_call",
                                                        module: "module/test_operator_lowering_lowers_plus_symbol_into_function_call",
                                                        start: 54,
                                                        end: 95,
                                                    },
                                                    statements: StatementList {
                                                        statements: [
                                                            Return(
                                                                ReturnStatement {
                                                                    span: Span {
                                                                        file: "test_operator_lowering_lowers_plus_symbol_into_function_call",
                                                                        module: "module/test_operator_lowering_lowers_plus_symbol_into_function_call",
                                                                        start: 72,
                                                                        end: 81,
                                                                    },
                                                                    expression: Literal(
                                                                        Integer(
                                                                            IntegerLiteral {
                                                                                span: Span {
                                                                                    file: "test_operator_lowering_lowers_plus_symbol_into_function_call",
                                                                                    module: "module/test_operator_lowering_lowers_plus_symbol_into_function_call",
                                                                                    start: 79,
                                                                                    end: 80,
                                                                                },
                                                                                value: 0,
                                                                            },
                                                                        ),
                                                                    ),
                                                                },
                                                            ),
                                                        ],
                                                    },
                                                },
                                            },
                                        ),
                                    ],
                                },
                            },
                        ),
                        FunctionDeclaration(
                            FunctionDeclarationElement {
                                span: Span {
                                    file: "test_operator_lowering_lowers_plus_symbol_into_function_call",
                                    module: "module/test_operator_lowering_lowers_plus_symbol_into_function_call",
                                    start: 115,
                                    end: 252,
                                },
                                decorators: FunctionDecorators {
                                    annotations: [],
                                    export: false,
                                },
                                ident: Identifier {
                                    span: Span {
                                        file: "test_operator_lowering_lowers_plus_symbol_into_function_call",
                                        module: "module/test_operator_lowering_lowers_plus_symbol_into_function_call",
                                        start: 124,
                                        end: 128,
                                    },
                                    value: "main",
                                },
                                params: FormalParameterList {
                                    span: Span {
                                        file: "test_operator_lowering_lowers_plus_symbol_into_function_call",
                                        module: "module/test_operator_lowering_lowers_plus_symbol_into_function_call",
                                        start: 128,
                                        end: 130,
                                    },
                                    parameters: [],
                                },
                                returns: Some(
                                    TypeAnnotation {
                                        span: Span {
                                            file: "test_operator_lowering_lowers_plus_symbol_into_function_call",
                                            module: "module/test_operator_lowering_lowers_plus_symbol_into_function_call",
                                            start: 132,
                                            end: 135,
                                        },
                                        ty: Binding(
                                            "i32",
                                        ),
                                    },
                                ),
                                body: BlockStatement {
                                    span: Span {
                                        file: "test_operator_lowering_lowers_plus_symbol_into_function_call",
                                        module: "module/test_operator_lowering_lowers_plus_symbol_into_function_call",
                                        start: 136,
                                        end: 252,
                                    },
                                    statements: StatementList {
                                        statements: [
                                            Variable(
                                                VariableStatement {
                                                    span: Span {
                                                        file: "test_operator_lowering_lowers_plus_symbol_into_function_call",
                                                        module: "module/test_operator_lowering_lowers_plus_symbol_into_function_call",
                                                        start: 150,
                                                        end: 165,
                                                    },
                                                    modifier: Let(
                                                        Span {
                                                            file: "test_operator_lowering_lowers_plus_symbol_into_function_call",
                                                            module: "module/test_operator_lowering_lowers_plus_symbol_into_function_call",
                                                            start: 150,
                                                            end: 153,
                                                        },
                                                    ),
                                                    target: Identifier(
                                                        Identifier {
                                                            span: Span {
                                                                file: "test_operator_lowering_lowers_plus_symbol_into_function_call",
                                                                module: "module/test_operator_lowering_lowers_plus_symbol_into_function_call",
                                                                start: 154,
                                                                end: 155,
                                                            },
                                                            value: "a",
                                                        },
                                                    ),
                                                    expression: Literal(
                                                        Integer(
                                                            IntegerLiteral {
                                                                span: Span {
                                                                    file: "test_operator_lowering_lowers_plus_symbol_into_function_call",
                                                                    module: "module/test_operator_lowering_lowers_plus_symbol_into_function_call",
                                                                    start: 163,
                                                                    end: 164,
                                                                },
                                                                value: 1,
                                                            },
                                                        ),
                                                    ),
                                                    type_annotation: Some(
                                                        TypeAnnotation {
                                                            span: Span {
                                                                file: "test_operator_lowering_lowers_plus_symbol_into_function_call",
                                                                module: "module/test_operator_lowering_lowers_plus_symbol_into_function_call",
                                                                start: 157,
                                                                end: 160,
                                                            },
                                                            ty: Binding(
                                                                "i32",
                                                            ),
                                                        },
                                                    ),
                                                },
                                            ),
                                            Variable(
                                                VariableStatement {
                                                    span: Span {
                                                        file: "test_operator_lowering_lowers_plus_symbol_into_function_call",
                                                        module: "module/test_operator_lowering_lowers_plus_symbol_into_function_call",
                                                        start: 178,
                                                        end: 193,
                                                    },
                                                    modifier: Let(
                                                        Span {
                                                            file: "test_operator_lowering_lowers_plus_symbol_into_function_call",
                                                            module: "module/test_operator_lowering_lowers_plus_symbol_into_function_call",
                                                            start: 178,
                                                            end: 181,
                                                        },
                                                    ),
                                                    target: Identifier(
                                                        Identifier {
                                                            span: Span {
                                                                file: "test_operator_lowering_lowers_plus_symbol_into_function_call",
                                                                module: "module/test_operator_lowering_lowers_plus_symbol_into_function_call",
                                                                start: 182,
                                                                end: 183,
                                                            },
                                                            value: "b",
                                                        },
                                                    ),
                                                    expression: Literal(
                                                        Integer(
                                                            IntegerLiteral {
                                                                span: Span {
                                                                    file: "test_operator_lowering_lowers_plus_symbol_into_function_call",
                                                                    module: "module/test_operator_lowering_lowers_plus_symbol_into_function_call",
                                                                    start: 191,
                                                                    end: 192,
                                                                },
                                                                value: 2,
                                                            },
                                                        ),
                                                    ),
                                                    type_annotation: Some(
                                                        TypeAnnotation {
                                                            span: Span {
                                                                file: "test_operator_lowering_lowers_plus_symbol_into_function_call",
                                                                module: "module/test_operator_lowering_lowers_plus_symbol_into_function_call",
                                                                start: 185,
                                                                end: 188,
                                                            },
                                                            ty: Binding(
                                                                "i32",
                                                            ),
                                                        },
                                                    ),
                                                },
                                            ),
                                            Variable(
                                                VariableStatement {
                                                    span: Span {
                                                        file: "test_operator_lowering_lowers_plus_symbol_into_function_call",
                                                        module: "module/test_operator_lowering_lowers_plus_symbol_into_function_call",
                                                        start: 206,
                                                        end: 220,
                                                    },
                                                    modifier: Let(
                                                        Span {
                                                            file: "test_operator_lowering_lowers_plus_symbol_into_function_call",
                                                            module: "module/test_operator_lowering_lowers_plus_symbol_into_function_call",
                                                            start: 206,
                                                            end: 209,
                                                        },
                                                    ),
                                                    target: Identifier(
                                                        Identifier {
                                                            span: Span {
                                                                file: "test_operator_lowering_lowers_plus_symbol_into_function_call",
                                                                module: "module/test_operator_lowering_lowers_plus_symbol_into_function_call",
                                                                start: 210,
                                                                end: 211,
                                                            },
                                                            value: "x",
                                                        },
                                                    ),
                                                    expression: Arguments(
                                                        ArgumentsExpression {
                                                            span: Span {
                                                                file: "test_operator_lowering_lowers_plus_symbol_into_function_call",
                                                                module: "module/test_operator_lowering_lowers_plus_symbol_into_function_call",
                                                                start: 214,
                                                                end: 219,
                                                            },
                                                            ident: MemberDot(
                                                                MemberDotExpression {
                                                                    span: Span {
                                                                        file: "test_operator_lowering_lowers_plus_symbol_into_function_call",
                                                                        module: "module/test_operator_lowering_lowers_plus_symbol_into_function_call",
                                                                        start: 214,
                                                                        end: 219,
                                                                    },
                                                                    target: Identifier(
                                                                        IdentifierExpression {
                                                                            span: Span {
                                                                                file: "test_operator_lowering_lowers_plus_symbol_into_function_call",
                                                                                module: "module/test_operator_lowering_lowers_plus_symbol_into_function_call",
                                                                                start: 214,
                                                                                end: 215,
                                                                            },
                                                                            ident: Identifier {
                                                                                span: Span {
                                                                                    file: "test_operator_lowering_lowers_plus_symbol_into_function_call",
                                                                                    module: "module/test_operator_lowering_lowers_plus_symbol_into_function_call",
                                                                                    start: 214,
                                                                                    end: 215,
                                                                                },
                                                                                value: "a",
                                                                            },
                                                                        },
                                                                    ),
                                                                    expression: Identifier(
                                                                        IdentifierExpression {
                                                                            span: Span {
                                                                                file: "test_operator_lowering_lowers_plus_symbol_into_function_call",
                                                                                module: "module/test_operator_lowering_lowers_plus_symbol_into_function_call",
                                                                                start: 214,
                                                                                end: 219,
                                                                            },
                                                                            ident: Identifier {
                                                                                span: Span {
                                                                                    file: "test_operator_lowering_lowers_plus_symbol_into_function_call",
                                                                                    module: "module/test_operator_lowering_lowers_plus_symbol_into_function_call",
                                                                                    start: 214,
                                                                                    end: 219,
                                                                                },
                                                                                value: "plus",
                                                                            },
                                                                        },
                                                                    ),
                                                                },
                                                            ),
                                                            arguments: ArgumentsList {
                                                                span: Span {
                                                                    file: "test_operator_lowering_lowers_plus_symbol_into_function_call",
                                                                    module: "module/test_operator_lowering_lowers_plus_symbol_into_function_call",
                                                                    start: 214,
                                                                    end: 219,
                                                                },
                                                                arguments: [
                                                                    Identifier(
                                                                        IdentifierExpression {
                                                                            span: Span {
                                                                                file: "test_operator_lowering_lowers_plus_symbol_into_function_call",
                                                                                module: "module/test_operator_lowering_lowers_plus_symbol_into_function_call",
                                                                                start: 218,
                                                                                end: 219,
                                                                            },
                                                                            ident: Identifier {
                                                                                span: Span {
                                                                                    file: "test_operator_lowering_lowers_plus_symbol_into_function_call",
                                                                                    module: "module/test_operator_lowering_lowers_plus_symbol_into_function_call",
                                                                                    start: 218,
                                                                                    end: 219,
                                                                                },
                                                                                value: "b",
                                                                            },
                                                                        },
                                                                    ),
                                                                ],
                                                            },
                                                        },
                                                    ),
                                                    type_annotation: None,
                                                },
                                            ),
                                            Return(
                                                ReturnStatement {
                                                    span: Span {
                                                        file: "test_operator_lowering_lowers_plus_symbol_into_function_call",
                                                        module: "module/test_operator_lowering_lowers_plus_symbol_into_function_call",
                                                        start: 240,
                                                        end: 241,
                                                    },
                                                    expression: Literal(
                                                        Integer(
                                                            IntegerLiteral {
                                                                span: Span {
                                                                    file: "test_operator_lowering_lowers_plus_symbol_into_function_call",
                                                                    module: "module/test_operator_lowering_lowers_plus_symbol_into_function_call",
                                                                    start: 240,
                                                                    end: 241,
                                                                },
                                                                value: 0,
                                                            },
                                                        ),
                                                    ),
                                                },
                                            ),
                                        ],
                                    },
                                },
                            },
                        ),
                    ],
                },
            },
        ],
    },
}
